contracts/core/src/contract.rs

```diff
@@ -346,6 +369,16 @@ fn execute_remove_bond_provider(
     cw_ownable::assert_owner(deps.storage, &info.sender)?;
 
     let bond_provider_address = deps.api.addr_validate(&bond_provider_address)?;
+    let bond_provider_balances = deps
+        .querier
+        .query_all_balances(bond_provider_address.clone())?;
+    let bond_provider_balances_except_untrn = bond_provider_balances
+        .into_iter()
+        .filter(|coin| coin.denom != *UNTRN_DENOM.to_string())
+        .collect::<Vec<Coin>>();
+    if !bond_provider_balances_except_untrn.is_empty() {
+        return Err(ContractError::BondProviderBalanceNotEmpty {});
+    }
```

Comment:
I think we should add query interface to the bond provider and let him answer can it be removed or not. Because there is internal state in the provider and it relates not only to bank balance but on other operations as well. So the real way to remove provider on working protocol is to pause bonding, wait for all balances and internal states becomes normal and only after that we can safely remove bond provider

Reply:
I agree that it's much more convenitent to add such a query; nevertheless, I propose to keep these errors in execution methods